sql/scan.go

```diff
@@ -231,22 +235,39 @@ func (n *scanNode) initScan() bool {
 		return true
 	}
 
-	// Retrieve all of the keys that start with our index key prefix.
-	if len(n.startKey) == 0 {
-		n.startKey = proto.Key(MakeIndexKeyPrefix(n.desc.ID, n.index.ID))
-	}
-	if len(n.endKey) == 0 {
-		n.endKey = n.startKey.PrefixEnd()
+	if len(n.spans) == 0 {
+		// If no spans were specified retrieve all of the keys that start with our
+		// index key prefix.
+		start := proto.Key(MakeIndexKeyPrefix(n.desc.ID, n.index.ID))
+		n.spans = append(n.spans, span{
+			start: start,
+			end:   start.PrefixEnd(),
+		})
 	}
+
+	// Retrieve all the spans.
+	b := &client.Batch{}
```

Comment:
In a case like `WHERE a IN (1,2,3,4,5,6,7) LIMIT 1` batching all the span reads will guarantee that the average cost of the query is also the worst-case cost. We'll want to do something more clever by reading `N` spans at a time where `N` is some derivative of the `LIMIT` number.